File: desktop/src/homemade/game/model/combo/ComboPack.java
package homemade.game.model.combo;

import homemade.game.Combo;
import homemade.game.fieldstructure.CellCode;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

/**
 * Represents a set of combos, created "simultaneously".
 * <p>
 * Must be synchronized externally!
 */
public class ComboPack {
    private ArrayList<Combo> combos;

    private int tier;
    private int bonusMultiplier;

    public ComboPack() {
        combos = new ArrayList<>();
        tier = 0;
        bonusMultiplier = 0;
    }

    void addMultiplier(int value) {
        bonusMultiplier += value;
    }

    void add(Combo combo) {
        combos.add(combo);
        tier += combo.getTier();
    }

    public void append(ComboPack anotherPack) {
        combos.addAll(anotherPack.combos);
        tier += anotherPack.tier;
        bonusMultiplier += anotherPack.bonusMultiplier;
    }

    public Set<CellCode> cellSet() {
        Set<CellCode> cells = new HashSet<>();

        for (Combo combo : combos)
            cells.addAll(combo.toSet());

        return cells;
    }

    public Iterator<Combo> comboIterator() {
        return combos.iterator();
    }

    public int numberOfCombos() {
        return combos.size();
    }

    public int packTier() {
        return tier;
    }

    public int multiplierIncrease() {
        return tier + bonusMultiplier;
    }
}

File: game-model-v2/src/main/java/homemade/game/model/combo/ComboPack.kt
package homemade.game.model.combo

import homemade.game.fieldstructure.CellCode

/**
 * Represents a set of combos, created "simultaneously".
 */
class ComboPack {
    private val combos: ArrayList<Combo> = ArrayList()
    private var tier = 0
    private var bonusMultiplier = 0

    fun addMultiplier(value: Int) {
        bonusMultiplier += value
    }

    fun add(combo: Combo) {
        combos.add(combo)
        tier += combo.tier
    }

    fun append(anotherPack: ComboPack) {
        combos.addAll(anotherPack.combos)
        tier += anotherPack.tier
        bonusMultiplier += anotherPack.bonusMultiplier
    }

    fun cellSet(): Set<CellCode> {
        val cells: MutableSet<CellCode> = HashSet()
        for (combo in combos) cells.addAll(combo.toSet())
        return cells
    }

    fun comboIterator(): Iterator<Combo> {
        return combos.iterator()
    }

    fun numberOfCombos(): Int {
        return combos.size
    }

    fun packTier(): Int {
        return tier
    }

    fun multiplierIncrease(): Int {
        return tier + bonusMultiplier
    }
}
